### ISSUE TITLE
instanceof / BC_IMPOSSIBLE_INSTANCEOF false positive in catch block

### ISSUE THREAD

**Issue by rjclark** [spotbugs/spotbugs#926]
I'm seeing what I think is false positive in this code block

```java
try {
   ...
}
catch (Exception e) {
    if ( e instanceof InterruptedException ) { // <- false positive
        ...
    }
}
```

> instanceof will always return false
> This instanceof test will always return false. Although this is safe, make sure it isn't an indication of some misunderstanding or some other logic error.
>
> Bug kind and pattern: BC - BC_IMPOSSIBLE_INSTANCEOF

>  instanceof will always return false in [...], since a RuntimeException can't be a InterruptedException

It appears as through the `Exception` is being turned into a `RuntimeException` incorrectly, before analysis.

This also appears to have been reported against Findbugs:

https://sourceforge.net/p/findbugs/bugs/1463/

I've seen the same problem with IntelliJ's inspections

**Comment by Xiaoven** [comment]
Why not use `catch (InterruptedException){ ...}`?

**Comment by hazendaz** [comment]
Old and closing but commenting for others.

I don't see this as a false positive, it flagged your code, the code is in fact incorrectly used.  Catch exceptions correctly.  Do not use instanceof within exceptions.  As noted you said it was a runtimeException.  Well...then it really was a runtimeException not an interrupted exception, they are not the same.  Either way as @Xiaoven noted, write correctly.  I've never seen such usage in 15 years coding java an instanceof used that way.  Getting into checked vs unchecked exceptions here and while spotbugs may not have directly noted exactly the issue, its still flagging an issue which is fairly obvious to the trained eye on exception processing.

**Comment by PVoLan** [comment]
Hi @hazendaz 

I've just faced similar issue and would insist that it is a false-positive.

The following Kotlin code

```
fun main(){
    try {
        foo()
    } catch (e: Exception){
        extracted(e)
    }
}

private fun extracted(e: Exception) {
    println("Error! General information and general handling")
    if (e is SpecialException) {
        println("More specific error details and specific handling: ${e.errorCode}")
    }
}

private fun foo(){
    throw SpecialException(15)
}

private class SpecialException(val errorCode: Int) : Exception()
```

triggers Spotbug warning
```
Code | Warning
BC | instanceof will always return false in FooKt.main(), since a RuntimeException cannot be a SpecialException
```

- First of all, this is simply misleading. We have no any usage of RuntimeException here (it's valid for Kotlin). If you don't like use of is/instance of with exceptions, you should write it directly in the warning message - do not drag RuntimeException into dicsussion
- Second, the sample above is a very typical way of handling exceptions. I have one general catch, because I have a very general handler for all the errors possible. But for some specific exceptions I want more specific handling, and this is where using of is/instance of is completely reasonable
